grpc/shared_conn.go

```diff
@@ -18,68 +24,221 @@ type OnTrackCB func(tr *webrtc.TrackRemote, r *webrtc.RTPReceiver)
+	if err != nil {
+		sc.logger.Warnw("Unable to create optional peer connection for module. Ignoring.", "err", err)
+		return
+	}
+
+	sc.peerConn = peerConn
+	sc.peerConnReady, err = rpc.ConfigureForRenegotiation(peerConn, sc.logger.AsZap())
+	if err != nil {
+		sc.logger.Warnw("Unable to create optional renegotiation channel for module. Ignoring.", "err", err)
+		utils.UncheckedError(sc.peerConn.Close())
+		sc.peerConn = nil
+		return
+	}
+
+	sc.peerConn.OnTrack(func(trackRemote *webrtc.TrackRemote, rtpReceiver *webrtc.RTPReceiver) {
+		name, err := resource.NewFromString(trackRemote.StreamID())
+		if err != nil {
+			sc.logger.Errorw("StreamID did not parse as a ResourceName", "sharedConn", fmt.Sprintf("%p", sc), "streamID", trackRemote.StreamID())
+			return
+		}
+		sc.resOnTrackMu.Lock()
+		onTrackCB, ok := sc.resOnTrackCBs[name]
+		sc.resOnTrackMu.Unlock()
+		if !ok {
+			sc.logger.Errorw("Callback not found for StreamID", "sharedConn", fmt.Sprintf("%p", sc), "streamID", trackRemote.StreamID())
+			return
+		}
+		onTrackCB(trackRemote, rtpReceiver)
+	})
+}
+
+// GenerateEncodedOffer creates a WebRTC offer that's JSON + base64 encoded. If an error is
+// returned, `SharedConn.PeerConn` will return nil until a following `Reset`.
+func (sc *SharedConn) GenerateEncodedOffer() (string, error) {
+	success := false
+	defer func() {
+		// Only keep the `peerConn` pointer active if we believe the connection is useable.
+		if !success {
+			sc.peerConn = nil
+		}
+	}()
+
+	pc := sc.PeerConn()
+	if pc == nil {
+		return "", errors.New("peer connections disabled")
+	}
+
+	offer, err := pc.CreateOffer(nil)
+	if err != nil {
+		return "", err
+	}
+
+	if err = pc.SetLocalDescription(offer); err != nil {
+		return "", err
+	}
+
+	<-webrtc.GatheringCompletePromise(pc)
+	success = true
+	return rpc.EncodeSDP(pc.LocalDescription())
+}
+
+// ProcessEncodedAnswer sets the remote description to the answer and waits for the connection to be
+// ready as per the negotiation channel being opened. The answer is expected to be JSON + base64
+// encoded. If an error is returned, `SharedConn.PeerConn` will return nil until a following
+// `Reset`.
+func (sc *SharedConn) ProcessEncodedAnswer(encodedAnswer string) error {
+	success := false
+	defer func() {
+		// Only keep the `peerConn` pointer active if we believe the connection is useable.
+		if !success {
+			sc.peerConn = nil
+		}
+	}()
+
+	pc := sc.PeerConn()
+	if pc == nil {
+		return errors.New("PeerConnection was not initialized")
+	}
+
+	answer := webrtc.SessionDescription{}
+	if err := rpc.DecodeSDP(encodedAnswer, &answer); err != nil {
+		return err
+	}
+
+	if err := pc.SetRemoteDescription(answer); err != nil {
+		return err
+	}
+
+	<-sc.peerConnReady
+	success = true
+	return nil
+}
+
 // Close closes a shared connection.
 func (sc *SharedConn) Close() error {
 	var err error
 	if sc.peerConn != nil {
 		err = sc.peerConn.Close()
+		sc.peerConn = nil
```

Comment:
If we are serializing access to `sc.peerConn` Do we also need to hold the mutex during before & during this if statement?